src/sha256.nr

```diff
@@ -153,6 +153,19 @@ pub(crate) fn process_full_blocks<let N: u32>(
         states[i + 1] = sha256_compression(new_msg_block, states[i]);
     }
 
+    // If message_size/BLOCK_SIZE == N/BLOCK_SIZE, and there is a remainder, we need to process the last block.
+    if N % BLOCK_SIZE != 0 {
+        let new_msg_block =
+            // Safety: separate verification function
+            unsafe { build_msg_block(msg, message_size, BLOCK_SIZE * num_blocks) };
+
+        // Verify the block we are compressing was appropriately constructed
+        verify_msg_block(msg, message_size, new_msg_block, BLOCK_SIZE * num_blocks);
+        verify_msg_block_zeros(new_msg_block, message_size % BLOCK_SIZE, INT_BLOCK_SIZE);
```

Comment:
I think that merging this into a `verify_partial_block` will reduce constraints even more.

`verify_msg_block` does a conditional constraint up to `message_size` and `verify_msg_block_zeros` has a conditional constraint that everything after `message_size` is zero but these run over the full length of the block. We could instead have an unconditional constraint which constrains the value to be either matching the value calculated from `msg` or zero.

```noir
let expected_value = if i > message_size { /* block value */} else {0};
assert_eq(msg_block[i], expected_value);
```

Reply:
Just got to make sure that this doesn't interfere with the full block verification.